docs/developers/napari_models.md

```diff
@@ -0,0 +1,226 @@
+# Napari models and events
+
+The napari codebase can be thought to consist of three main components:
+
+* python models describing objects - these are able to operate separately from
+  the viewer and do not have any dependencies on user interface classes
+    * this code lives in `napari/components` (utility objects) and
+     `napari/layers` (objects that contain data)
+* qt classes that handle the interactive GUI aspect of the napari viewer
+    * the private qt code lives in `napari/_qt` and the smaller public qt
+      interface code lives in `napari/qt`
+* vispy classes that handle rendering
+    * the code for this is private and lives in `napari/_vispy`
+
+The separation of the python models from viewer GUI code allows:
+
+* the python model to be easily run headless without the viewer, for example
```

Comment:
I realise don't actually know how to use napari in headless mode.... :grimacing:

Reply:
I couldn't find anything in the docs, maybe it would also be a good addition to the documentation (maybe a page in the 'getting started' section..)

Reply:
i'd say it's more of a "concept" at this point :joy: ... with probably not a huge amount of use cases.  But it's something that we generally protect (in terms of keeping the gui isolated) out of hopes for more future usage.

Headless mode is pretty much just anything without showing the viewer
```python
viewer = napari.Viewer(show=False)
```

but one could also directly use lower level components like the `LayerList` and `Layer` objects (without a viewer).

I'll note that currently, using `Viewer(show=False)` still won't prevent qt stuff from being imported... however, if one uninstalls qtpy or any of the qt components, then it will work just fine (and qt won't be initialized).

this is all something that's good to know for a napari dev, but I'm not sure it's something that someone getting started with napari would need to understand (or should be exposed to just yet).

Reply:
Yeah agreed not necessarily needed for a user but since this is developer facings docs, I'd say it's worth even just summarising @tlambert03's comment above in a little admonition box

Reply:
Thanks! I will update the docs with the info